File: src/CSharp/CodeCracker/Style/TernaryOperatorAnalyzer.cs
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.Diagnostics;
using System.Collections.Immutable;
using System.Linq;

namespace CodeCracker.CSharp.Style
{
    [DiagnosticAnalyzer(LanguageNames.CSharp)]
    public class TernaryOperatorAnalyzer : DiagnosticAnalyzer
    {
        internal const string Title = "Use ternary operator";
        internal const string MessageFormatForIfWithReturn = "{0}";
        internal const string Category = SupportedCategories.Style;
        internal const string MessageFormatForIfWithAssignment = "{0}";

        internal static readonly DiagnosticDescriptor RuleForIfWithReturn = new DiagnosticDescriptor(
            DiagnosticId.TernaryOperator_Return.ToDiagnosticId(),
            Title,
            MessageFormatForIfWithReturn,
            Category,
            DiagnosticSeverity.Info,
            isEnabledByDefault: true,
            helpLinkUri: HelpLink.ForDiagnostic(DiagnosticId.TernaryOperator_Return));

        internal static readonly DiagnosticDescriptor RuleForIfWithAssignment = new DiagnosticDescriptor(
            DiagnosticId.TernaryOperator_Assignment.ToDiagnosticId(),
            Title,
            MessageFormatForIfWithAssignment,
            Category,
            DiagnosticSeverity.Info,
            isEnabledByDefault: true,
            helpLinkUri: HelpLink.ForDiagnostic(DiagnosticId.TernaryOperator_Assignment));

        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(RuleForIfWithReturn, RuleForIfWithAssignment);

        public override void Initialize(AnalysisContext context) =>
            context.RegisterSyntaxNodeAction(Analyzer, SyntaxKind.IfStatement);

        private static void Analyzer(SyntaxNodeAnalysisContext context)
        {
            if (context.IsGenerated()) return;
            var ifStatement = context.Node as IfStatementSyntax;
            if (ifStatement == null) return;
            if (ifStatement.Else == null) return;
            var blockIf = ifStatement.Statement as BlockSyntax;
            var blockElse = ifStatement.Else.Statement as BlockSyntax;
            if ((blockIf == null || blockIf.Statements.Count == 1) &&
                (blockElse == null || blockElse.Statements.Count == 1))
            {
                var statementInsideIf = ifStatement.Statement is BlockSyntax ? ((BlockSyntax)ifStatement.Statement).Statements.Single() : ifStatement.Statement;
                var elseStatement = ifStatement.Else;
                var statementInsideElse = elseStatement.Statement is BlockSyntax ? ((BlockSyntax)elseStatement.Statement).Statements.Single() : elseStatement.Statement;
                if (statementInsideIf is ReturnStatementSyntax && statementInsideElse is ReturnStatementSyntax)
                {
                    var diagnostic = Diagnostic.Create(RuleForIfWithReturn, ifStatement.IfKeyword.GetLocation(), "You can use a ternary operator.");
                    context.ReportDiagnostic(diagnostic);
                    return;
                }
                var expressionInsideIf = statementInsideIf as ExpressionStatementSyntax;
                var expressionInsideElse = statementInsideElse as ExpressionStatementSyntax;
                if (expressionInsideIf != null && expressionInsideIf.Expression.IsKind(SyntaxKind.SimpleAssignmentExpression) && expressionInsideElse != null && expressionInsideElse.Expression.IsKind(SyntaxKind.SimpleAssignmentExpression))
                {
                    var assignmentExpressionInsideIf = (AssignmentExpressionSyntax)expressionInsideIf.Expression;
                    var assignmentExpressionInsideElse = (AssignmentExpressionSyntax)expressionInsideElse.Expression;
                    var variableIdentifierInsideIf = assignmentExpressionInsideIf.Left as IdentifierNameSyntax;
                    var variableIdentifierInsideElse = assignmentExpressionInsideElse.Left as IdentifierNameSyntax;
                    if (variableIdentifierInsideIf == null || variableIdentifierInsideElse == null
                        || variableIdentifierInsideIf.Identifier.Text != variableIdentifierInsideElse.Identifier.Text) return;
                    var diagnostic = Diagnostic.Create(RuleForIfWithAssignment, ifStatement.IfKeyword.GetLocation(), "You can use a ternary operator.");
                    context.ReportDiagnostic(diagnostic);
                }
            }
        }
    }
}

File: src/VisualBasic/CodeCracker/Style/TernaryOperatorAnalyzer.vb
Imports System.Collections.Immutable
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.Diagnostics
Imports Microsoft.CodeAnalysis.VisualBasic
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax

Namespace Style
    <DiagnosticAnalyzer(LanguageNames.VisualBasic)>
    Public Class TernaryOperatorAnalyzer
        Inherits DiagnosticAnalyzer

        Friend Const Title = "Use Ternary operator."
        Friend Const MessageFormat = "You can use a ternary operator."

        Friend Shared RuleForIfWithReturn As New DiagnosticDescriptor(
            DiagnosticId.TernaryOperator_Return.ToDiagnosticId(),
            Title,
            MessageFormat,
            SupportedCategories.Style,
            DiagnosticSeverity.Warning,
            isEnabledByDefault:=True,
            helpLinkUri:=HelpLink.ForDiagnostic(DiagnosticId.TernaryOperator_Return))

        Friend Shared RuleForIfWithAssignment As New DiagnosticDescriptor(
            DiagnosticId.TernaryOperator_Assignment.ToDiagnosticId(),
            Title,
            MessageFormat,
            SupportedCategories.Style,
            DiagnosticSeverity.Warning,
            isEnabledByDefault:=True,
            helpLinkUri:=HelpLink.ForDiagnostic(DiagnosticId.TernaryOperator_Assignment))

        Friend Shared RuleForIif As New DiagnosticDescriptor(
            DiagnosticId.TernaryOperator_Iif.ToDiagnosticId(),
            Title,
            MessageFormat,
            SupportedCategories.Style,
            DiagnosticSeverity.Warning,
            isEnabledByDefault:=True,
            helpLinkUri:=HelpLink.ForDiagnostic(DiagnosticId.TernaryOperator_Iif))

        Public Overrides ReadOnly Property SupportedDiagnostics As ImmutableArray(Of DiagnosticDescriptor)
            Get
                Return ImmutableArray.Create(RuleForIfWithAssignment, RuleForIfWithReturn, RuleForIif)
            End Get
        End Property

        Public Overrides Sub Initialize(context As AnalysisContext)
            context.RegisterSyntaxNodeAction(AddressOf Analyzer, SyntaxKind.IfStatement)
            context.RegisterSyntaxNodeAction(AddressOf IifAnalyzer, SyntaxKind.InvocationExpression)
        End Sub
        Private Sub Analyzer(context As SyntaxNodeAnalysisContext)
            If (context.IsGenerated()) Then Return
            Dim ifStatement = TryCast(context.Node, IfStatementSyntax)
            If ifStatement Is Nothing Then Exit Sub
            Dim ifBlock = TryCast(ifStatement.Parent, MultiLineIfBlockSyntax)
            If ifBlock Is Nothing Then Exit Sub
            If ifBlock.ElseBlock Is Nothing Then Exit Sub
            If ifBlock.Statements.Count <> 1 OrElse ifBlock.ElseBlock.Statements.Count <> 1 Then Exit Sub

            Dim ifClauseStatement = ifBlock.Statements(0)
            Dim elseStatement = ifBlock.ElseBlock.Statements(0)

            If TypeOf (ifClauseStatement) Is ReturnStatementSyntax AndAlso
            TypeOf (elseStatement) Is ReturnStatementSyntax Then

                Dim diag = Diagnostic.Create(RuleForIfWithReturn, ifStatement.IfKeyword.GetLocation, "You can use a ternary operator.")
                context.ReportDiagnostic(diag)
                Exit Sub
            End If

            Dim ifAssignment = TryCast(ifClauseStatement, AssignmentStatementSyntax)
            Dim elseAssignment = TryCast(elseStatement, AssignmentStatementSyntax)
            If ifAssignment Is Nothing OrElse elseAssignment Is Nothing Then Exit Sub
            If Not ifAssignment?.Left.IsEquivalentTo(elseAssignment?.Left) Then Exit Sub
            Dim assignDiag = Diagnostic.Create(RuleForIfWithAssignment, ifStatement.IfKeyword.GetLocation(), "You can use a ternary operator.")
            context.ReportDiagnostic(assignDiag)
        End Sub

        Private Sub IifAnalyzer(context As SyntaxNodeAnalysisContext)
            If (context.IsGenerated()) Then Return
            Dim iifStatement = TryCast(context.Node, InvocationExpressionSyntax)
            If iifStatement Is Nothing Then Exit Sub
            Dim iifExpression = TryCast(iifStatement.Expression, IdentifierNameSyntax)
            If iifExpression Is Nothing Then Exit Sub
            If String.Equals(iifExpression.Identifier.ValueText, "iif", StringComparison.OrdinalIgnoreCase) Then
                If iifStatement.ArgumentList.Arguments.Count = 3 Then
                    Dim diag = Diagnostic.Create(RuleForIif, iifStatement.GetLocation())
                    context.ReportDiagnostic(diag)
                End If
            End If
        End Sub
    End Class
End Namespace
